rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/event/RMInitialState.java

```diff
@@ -55,67 +58,118 @@
-        this.nodesList = nodesEventList;
-        this.nodeSources = nodeSourcesList;
+    public RMInitialState(Map<String, RMNodeEvent> nodesEventList, Map<String, RMNodeSourceEvent> nodeSourcesList) {
+        this.nodeEvents = nodesEventList;
+        this.nodeSourceEvents = nodeSourcesList;
     }
 
     /**
-     * Returns the 'to release' Nodes list.
-     * @return the 'to release' Nodes list.
+     * Current version of RM portal and maybe other clients expects "nodesEvents" inside JSON
+     *
+     * @return
      */
-    public ArrayList<RMNodeEvent> getNodesEvents() {
-        return this.nodesList;
+    public List<RMNodeEvent> getNodesEvents() {
+        return new ArrayList(this.nodeEvents.values());
     }
 
     /**
-     * Returns the NodeSources list.
-     * @return the NodeSources list.
+     * Current version of RM portal and maybe other clients expects "nodeSource" inside JSON
+     *
+     * @return
      */
-    public ArrayList<RMNodeSourceEvent> getNodeSource() {
-        return this.nodeSources;
+    public List<RMNodeSourceEvent> getNodeSource() {
+        return new ArrayList<>(this.nodeSourceEvents.values());
     }
 
-    public void nodeStateChanged(RMNodeEvent stateChangedEvent) {
-        int size = nodesList.size();
-        for (int i = 0; i < size; i++) {
-            if (stateChangedEvent.getNodeUrl().equals(nodesList.get(i).getNodeUrl())) {
-                nodesList.set(i, stateChangedEvent);
-                break;
-            }
-        }
+    public long getLatestCounter() {
+        return latestCounter;
+    }
+
+    public void nodeAdded(RMNodeEvent event) {
+        updateCounter(event);
+        nodeEvents.put(event.getNodeUrl(), event);
+
+    }
+
+    public void nodeStateChanged(RMNodeEvent event) {
+        updateCounter(event);
+        nodeEvents.put(event.getNodeUrl(), event);
     }
 
-    public void nodeRemoved(RMNodeEvent removedEvent) {
-        Iterator<RMNodeEvent> events = nodesList.iterator();
-        while (events.hasNext()) {
-            RMNodeEvent nodeEvent = events.next();
-            if (removedEvent.getNodeUrl().equals(nodeEvent.getNodeUrl())) {
-                events.remove();
-                break;
+    public void nodeRemoved(RMNodeEvent event) {
+        updateCounter(event);
+        nodeEvents.put(event.getNodeUrl(), event);
+    }
+
+    public void nodeSourceAdded(RMNodeSourceEvent event) {
+        updateCounter(event);
+        nodeSourceEvents.put(event.getSourceName(), event);
+    }
+
+    public void nodeSourceRemoved(RMNodeSourceEvent event) {
+        updateCounter(event);
+        nodeSourceEvents.remove(event.getSourceName());
+    }
+
+    private void updateCounter(RMEvent event) {
+        latestCounter = Math.max(latestCounter, event.getCounter());
+    }
+
+    public RMInitialState cloneAndFilter(long filter) {
+        long actualFilter = Math.min(filter, latestCounter);
```

Comment:
Hahaha) Real error was here.  
it should be:
long actualFilter = filter <= latestCouner ? filter : -1